### DISCUSSION TITLE
How train CGN to fill in missing node attributes?

### DISCUSSION THREAD

**Discussion by alexrutherford** [pyg-team/pytorch_geometric#3428] [Q&A] [answered]
Hi all

I would like to train a CGN to be able to predict attributes of new nodes added to my graph. So I am training on the nodes for which the attributes are known. 

I have a vector of N features (known) for each node, and the target is also one of those N features (the one I want to be able to predict for new nodes). I want to somehow 'hide' the known feature value of the node being predicted, so it is just predicted from the graph convolution of neighbours (and neighbours neighbours). Otherwise it is a very trivial prediction problem where the value of the feature of that node is predicted from that same value!

I think that setting `add_self_loops = False` in the final layer would do this: it means that the node v is not considered a neighbour of node v itself. Therefore the value I am trying to fit to in training would not be exposed.

Can anyone tell me if I got this right? Or does anyone have a good example of missing node feature prediction I could look at?

Many thanks!

**Comment by rusty1s** [comment, marked as answer]
I think you do not necessarily need to set `add_self_loops = False`, but rather change your training scenario as described [here](https://arxiv.org/abs/2009.03509). That is, in each epoch, you select a random number of training nodes, set their target feature to zero, and try to predict it.

**Reply by alexrutherford** [reply]
Thanks for the answer. I'm still a little confused

- Would `add_self_loops = False` have the effect that I want, equivalent to the masked training scenario from Shi et al?
- Wouldn't setting the target feature to zero be adding some (false) information?
- To randomly perturb the training data in each epoch, I would need to update the data and send it to the device each time through the loop. Are there any convenience functions to do that?

Many thanks for your help, I really appreciate it!

**Reply by rusty1s** [reply]
* For `GCNConv`, `add_self_loops=False` should work as well, although it limits the capabilities of GCN to preserve central node information.
* Not necessarily, as this false information is already injected in validation and tests nodes as well
* Your original `data.x` can stay on GPU, and you simply would perturb it similar to
```python
x = data.x.clone()
x[mask, feature_dim_to_set_to_zero] = 0.
out = GNN(x, data.edge_index)
```
